### ISSUE TITLE
Retry seems broken on subscription

### ISSUE THREAD

**Issue by akowald** [guidesmiths/rascal#26]
Hey there. I'm trying to follow the advanced example. I've made one of my handlers immediately throw an error which results in a channel error. My issue is that the retry option is ignored and so I am experiencing hundreds of restarting channels per second.
>    rascal:Subscription Received message: undefined from queue: import_service:project:import +3ms
  rascal:Subscription Handling channel error: crashed :( from import_project using channel: 2ecf933a-7575-4f01-8882-b8e47470352e +5ms
  app:log subscription.on('error') +9ms
  app:rascal crashed :( +8ms
  rascal:Subscription Subscribing to queue: import_service:project:import +0ms
  rascal:Vhost Requested channel. Outstanding channel requests: 1 +7ms
  rascal:SubscriberSession Removing channel: 2ecf933a-7575-4f01-8882-b8e47470352e from session +8ms
  rascal:Vhost Created channel 7879819d-ba7a-4995-b8da-869c69daa79f from connection: c736f742-5ad7-45f1-b7da-33c381788acc +2ms
\> Subscription.js: attachErrorHandlers
\> Subscription.js: channel.consume callback null { consumerTag: 'amq.ctag-v23v7N-_ej5WSBp4vQscmA' }

It seems the channel is restarted immediately regardless of retry's min property.. I've tried {retry:1000,retry:true,and leaving it out} My code is pretty similar to the advanced example, although for some reason in the example the errors aren't caught and the app stops.

Anyway, my question is, is this expected behavior to have the channel restarting so much or did I go wrong somewhere? Thanks for any advice and for this library!

Also: retry: { delay: 1000 } does nothing despite what the docs suggest. I think it was changed to retry: 1000? https://github.com/guidesmiths/rascal/issues/12

**Comment by cressie176** [comment]
Hi akowald,
Sorry you're having problems. I'll take a look and see if I can find out what's going on. It would help if you can share your code, either via a link to a github repo, gist or by pasting here.
Thanks,
Steve

**Comment by cressie176** [comment]
One thing it could be - are you listening for errors? As per the [very important section about event handling](https://github.com/guidesmiths/rascal#very-important-section-about-event-handling) 

> amqplib emits error events when a connection or channel encounters a problem. Rascal will listen for these and provided you use the default configuration will attempt automatic recovery (reconnection etc), however these events can indicate errors in your code, so it's also important to bring them to your attention. Rascal does this by re-emitting the error event, which means if you don't handle them, they will bubble up to the uncaught error handler and crash your application.

**Comment by cressie176** [comment]
OK - been looking into things a bit and I think I can clarify.

The subscription retry option is for automatically re-subscribing after a channel emits an error. The re-subscription is attempted immediately and if successful, and we're all good.

If re-subscription fails, then rascal will attempt to re-subscribe. This could lead to an infinite tight loop, which can be mitigated by the retry delay option.

In other words the retry delay option doesn't apply to channel errors, only re-subscription errors.

In your example I think the following is happening...

1. Subscriber receives a message
1. Subscriber throws an error, resulting in a channel error
1. The message is rolled back
1. Rascal re-subscribes successfully
1. GOTO 1

To resolve these types of error  catch the error in your handler and pass it to the `ackOrNack` with a [recovery strategy](https://github.com/guidesmiths/rascal/blob/master/README.md#message-acknowledgement-and-recovery-strategies), e.g.

```
subscription.on('message', (message, content, ackOrNack) => {
  doAsyncStuff(content, (err) => {
    if (err) return ackOrNack(err, { strategy: 'nack', defer: 1000, requeue: true })
    ackOrNack();
  })
}).on('error', (err) => console.error(err.message))
```

You'll need to vary the above depending on whether your code is synchronous, uses promises or async await.

**Comment by akowald** [comment]
Firstly, thanks for the reply. I should have provided code from the start: https://gist.github.com/akowald/2bbef80a69567e004d0a05231a9bfb82

It's similar to the example except that I commented out the redelivery option in the subscription config so that it wouldn't catch the message.

I've added defer to the recovery strategies like you suggested but I'm still seeing the channel spam above.  It will always emit a subscription error, triggering [debug('subscription.on(error)', err.message)](https://gist.github.com/akowald/2bbef80a69567e004d0a05231a9bfb82#file-index-js-L43) so I'm uncertain how I can use ackOrNack and use defer in this instance.
